### PR TITLE
Added `Functions` features and overall project rework

### PR DESCRIPTION
**Pull request by pinebit**
The main goal was to integrate @chainlink/contract package to avoid using hard-coded ABIs.
The other goal is to introduce a single Functions feature to evaluate feasibility.

### CODE REVIEW

File: src/HardhatChainlink.ts [hunk 1]
```diff
@@ -1,91 +1,93 @@
 import "@nomiclabs/hardhat-ethers/";
 import { BigNumber, BytesLike } from "ethers";
 import { HardhatRuntimeEnvironment } from "hardhat/types";
+
 import {
-  getLatestPrice,
```
Comment (pinebit) [inline src/HardhatChainlink.ts:5]: Lots of shuffling due to re-formatting the code with prettier...

File: package.json [hunk 2]
```diff
@@ -31,6 +31,7 @@
     "DOCUMENTATION.md"
   ],
   "devDependencies": {
+    "@chainlink/contracts": "0.6.1",
```
Comment (pinebit) [inline package.json:34]: Am I correct here? Probably this should not be a dev-dep, as long as postinstall needs to generate typings..

Comment (pinebit) [reply package.json]: Moved to `dependencies`.

**Comment by pinebit** [comment]
> Overall LGTM! I've added the documentation part.
> 
> I actually have one question: Will deleting the `contracts/` folder affect [hardhat tasks regarding running the Chainlink node](https://github.com/smartcontractkit/hardhat-chainlink/blob/main/DOCUMENTATION.md?rgh-link-date=2023-01-27T22%3A37%3A21Z#running-a-chainlink-node-on-a-hardhat-network)? Thanks in advance!

You are very right. I found the tasks are relying on two contracts: LinkToken and Oracle, so I added those back, but referenced from the official package. Also added a few contracts Functions will require.

**Comment by KuphJr** [comment]
I just finished reviewing the code and it looks good, however I need to pull this branch and test it out locally before I approve.  I should have time to do it tomorrow.

A few other TODO items that will need to be supported are:
- Off-chain secrets support
- A `AutoFunctionsConsumer.sol` contract that integrates with Automation
- Commands to deploy, fund & register `AutoFunctionsConsumer.sol` for upkeep

**Comment by pinebit** [comment]
> I just finished reviewing the code and it looks good, however I need to pull this branch and test it out locally before I approve. I should have time to do it tomorrow.
> 
> A few other TODO items that will need to be supported are:
> 
> * Off-chain secrets support
> * A `AutoFunctionsConsumer.sol` contract that integrates with Automation
> * Commands to deploy, fund & register `AutoFunctionsConsumer.sol` for upkeep

Even though this work is paused, I think we still need to move it forward (as is), because it addresses many other challenges in the plugin, not necessary related to Functions.

**Pull request merged**